asttokens/util.py

```diff
@@ -311,3 +316,39 @@ def get(self, obj, cls):
       method = getattr(obj, name, obj.visit_default)
       self._cache[cls] = method
     return method
+
+
+if sys.version_info[0] == 2:
+  # Python 2 doesn't support non-ASCII identifiers, and making the real patched_generate_tokens support Python 2
+  # means working with raw tuples instead of tokenize.TokenInfo namedtuples.
+  def patched_generate_tokens(original_tokens):
+    # type: (Any) -> Any
+    return original_tokens
+else:
+  def patched_generate_tokens(original_tokens):
+    # type: (Iterator[tokenize.TokenInfo]) -> Iterator[tokenize.TokenInfo]
+    """
+    Fixes tokens yielded by `tokenize.generate_tokens` to handle more non-ASCII characters in identifiers.
+    Workaround for https://github.com/python/cpython/issues/68382.
+    Should only be used when tokenizing a string that is known to be valid syntax,
+    because it assumes that error tokens are not actually errors.
+    Combines groups of consecutive NAME and/or ERRORTOKEN tokens into a single NAME token.
+    """
+    for key, group_iter in itertools.groupby(
+        original_tokens,
+        lambda t: tokenize.NAME if t.type == tokenize.ERRORTOKEN else t.type,
+    ):
+      group = list(group_iter)  # type: List[tokenize.TokenInfo]
+      if key == tokenize.NAME and len(group) > 1 and any(tok.type == tokenize.ERRORTOKEN for tok in group):
+        line = group[0].line  # type: str
+        assert {tok.line for tok in group} == {line}
```

Comment:
Should this check that the end of one token is the start of the next one in this group, in order to avoid merging tokens that are separated by whitespace?